### ISSUE TITLE
seq2seq.dynamic_decode with float16 inputs on CPU

### ISSUE THREAD

**Issue by georgesterpu** [tensorflow/tensorflow#22176]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Manjaro Linux, testing
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: nice
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.10.1
- **Python version**: 3.7.0
- **Bazel version (if compiling from source)**: 0.16.1
- **GCC/Compiler version (if compiling from source)**: 7.3.1
- **CUDA/cuDNN version**: cuda 9.2.148, cudnn 7.2.1
- **GPU model and memory**: gtx 1080
- **Exact command to reproduce**:
I am using my own setup here:
https://github.com/georgesterpu/Sigmedia-AVSR/blob/master/experiment_tcd_audio.py
If really needed, I could write a minimal example to reproduce the issue.

### Describe the problem
I have an issue with the code indicated above, when switching to 16 bit precision.
On the CPU only, the code crashes with the following error:

```
InvalidArgumentError (see above for traceback): indices[0] = 31 is not in [0, 31)
	 [[Node: Decoder/decoder/while/BasicDecoderStep/ScheduledEmbeddingTrainingHelperNextInputs/cond/embedding_lookup = GatherV2[Taxis=DT_INT32, Tindices=DT_INT32, Tparams=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"](Decoder/decoder/while/BasicDecoderStep/ScheduledEmbeddingTrainingHelperNextInputs/cond/embedding_lookup/Switch, Decoder/decoder/while/BasicDecoderStep/ScheduledEmbeddingTrainingHelperNextInputs/cond/GatherNd, Decoder/decoder/while/BasicDecoderStep/ScheduledEmbeddingTrainingHelperNextInputs/cond/embedding_lookup/axis)]]
``` 
This is supposed to be the embedding lookup of the input ids to a `seq2seq.BasicDecoder` taking a `seq2seq.ScheduledEmbeddingTrainingHelper`. I first created the embedding matrix in float16, then passed it to the `embedding` argument. As a workaround, I've also created it in float32, passing instead a callable that would cast the retrieved vectors to float16, but the error persists.

The number `31` is the size of the vocabulary, and no id is above 30. However, even if I increase the size of the embedding matrix, the error would look like this:
`indices[0] = 132 is not in [0, 132)`

The same code runs fine on GPU, but I am aware that error reporting is restricted in this case. 

Could you please help me find out where is the problem and how to fix it ?
Is `seq2seq.dynamic_decode` fully supporting float16 data types ?

Thank you

### Source code / logs
```
  File "/usr/lib/python3.7/site-packages/tensorflow/contrib/seq2seq/python/ops/decoder.py", line 322, in dynamic_decode
    swap_memory=swap_memory)
  File "/usr/lib/python3.7/site-packages/tensorflow/python/ops/control_flow_ops.py", line 3232, in while_loop
    return_same_structure)
  File "/usr/lib/python3.7/site-packages/tensorflow/python/ops/control_flow_ops.py", line 2952, in BuildLoop
    pred, body, original_loop_vars, loop_vars, shape_invariants)
  File "/usr/lib/python3.7/site-packages/tensorflow/python/ops/control_flow_ops.py", line 2887, in _BuildLoop
    body_result = body(*packed_vars_for_body)
  File "/usr/lib/python3.7/site-packages/tensorflow/contrib/seq2seq/python/ops/decoder.py", line 265, in body
    decoder_finished) = decoder.step(time, inputs, state)
  File "/usr/lib/python3.7/site-packages/tensorflow/contrib/seq2seq/python/ops/basic_decoder.py", line 146, in step
    sample_ids=sample_ids)
  File "/usr/lib/python3.7/site-packages/tensorflow/contrib/seq2seq/python/ops/helper.py", line 351, in next_inputs
    all_finished, lambda: base_next_inputs, maybe_sample)
  File "/usr/lib/python3.7/site-packages/tensorflow/python/util/deprecation.py", line 454, in new_func
    return func(*args, **kwargs)
  File "/usr/lib/python3.7/site-packages/tensorflow/python/ops/control_flow_ops.py", line 2057, in cond
    orig_res_f, res_f = context_f.BuildCondBranch(false_fn)
  File "/usr/lib/python3.7/site-packages/tensorflow/python/ops/control_flow_ops.py", line 1895, in BuildCondBranch
    original_result = fn()
  File "/usr/lib/python3.7/site-packages/tensorflow/contrib/seq2seq/python/ops/helper.py", line 340, in maybe_sample
    sampled_next_inputs = self._embedding_fn(sample_ids_sampling)
  File "/run/media/john_tukey/work/phd/65.cleanavsr/Sigmedia-AVSR/avsr/decoder_unimodal.py", line 454, in <lambda>
    embedding_fun = lambda ids: tf.cast(tf.nn.embedding_lookup(self._embedding_matrix, ids), dtype=self._hparams.dtype)
  File "/usr/lib/python3.7/site-packages/tensorflow/python/ops/embedding_ops.py", line 310, in embedding_lookup
    transform_fn=None)
  File "/usr/lib/python3.7/site-packages/tensorflow/python/ops/embedding_ops.py", line 133, in _embedding_lookup_and_transform
    result = _clip(array_ops.gather(params[0], ids, name=name),
  File "/usr/lib/python3.7/site-packages/tensorflow/python/ops/array_ops.py", line 2659, in gather
    return gen_array_ops.gather_v2(params, indices, axis, name=name)
  File "/usr/lib/python3.7/site-packages/tensorflow/python/ops/gen_array_ops.py", line 3142, in gather_v2
    "GatherV2", params=params, indices=indices, axis=axis, name=name)
  File "/usr/lib/python3.7/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/usr/lib/python3.7/site-packages/tensorflow/python/util/deprecation.py", line 454, in new_func
    return func(*args, **kwargs)
  File "/usr/lib/python3.7/site-packages/tensorflow/python/framework/ops.py", line 3155, in create_op
    op_def=op_def)
  File "/usr/lib/python3.7/site-packages/tensorflow/python/framework/ops.py", line 1717, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): indices[0] = 31 is not in [0, 31)
	 [[Node: Decoder/decoder/while/BasicDecoderStep/ScheduledEmbeddingTrainingHelperNextInputs/cond/embedding_lookup = GatherV2[Taxis=DT_INT32, Tindices=DT_INT32, Tparams=DT_HALF, _device="/job:localhost/replica:0/task:0/device:CPU:0"](Decoder/decoder/while/BasicDecoderStep/ScheduledEmbeddingTrainingHelperNextInputs/cond/embedding_lookup/Switch, Decoder/decoder/while/BasicDecoderStep/ScheduledEmbeddingTrainingHelperNextInputs/cond/GatherNd, Decoder/decoder/while/BasicDecoderStep/ScheduledEmbeddingTrainingHelperNextInputs/cond/embedding_lookup/axis)]]
```

**Comment by georgesterpu** [comment]
A similar issue in keras: https://github.com/keras-team/keras/issues/10531

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @asimshankar, @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @asimshankar, @ebrevdo: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @asimshankar, @ebrevdo: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
